features/FEATURE_BLE/ble/common/Duration.h

```diff
@@ -0,0 +1,578 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef BLE_COMMON_DURATION_H_
+#define BLE_COMMON_DURATION_H_
+
+#include <stdint.h>
+#include <stddef.h>
+#include "platform/mbed_assert.h"
+
+namespace ble {
+
+#if !defined(DOXYGEN_ONLY)
+
+/**
+ * Define a compile time range.
+ * @tparam Min left-bound
+ * @tparam Max right-bound
+ */
+template<uint32_t Min, uint32_t Max>
+struct Range {
+    static const uint32_t MIN = Min;
+    static const uint32_t MAX = Max;
+};
+
+/**
+ * Deduce default range for C++ basic integer types.
+ *
+ *
+ * @tparam Rep The C++ integer type.
+ */
+template<typename Rep>
+struct DefaultRange;
+
+/**
+ * DefaultRange specialization for uint8_t.
+ */
+template<>
+struct DefaultRange<uint8_t> {
+    typedef Range<0, 0xFF> type;
+};
+
+/**
+ * DefaultRange specialization for uint16_t.
+ */
+template<>
+struct DefaultRange<uint16_t > {
+    typedef Range<0, 0xFFFF> type;
+};
+
+/**
+ * DefaultRange specialization for uint32_t
+ */
+template<>
+struct DefaultRange<uint32_t> {
+    typedef Range<0, 0xFFFFFFFF> type;
+};
+
+/**
+ * Represent an integral compile time value that can be used in Duration.
+ *
+ * @tparam T Type of the integral value.
+ * @tparam V The integer value representing a never ending duration.
```

Comment:
Global note throughout the file: Please make sure "Duration" is capitalized when referring to the class but not otherwise.